Add MultiColumnsStackPanel

vgromfeld opened the pull request:
In one of my app, I have a very long form containing a lot of fields. I usually limit the width of the fields to have something not to big. To fill the rest of the screen, we can use adaptive visual state triggers but it is not that convenient.

In this PR, I'm proposing a multi columns stack panel that will automatically display its children in columns of equal height.
This control works best when used inside a vertical `ScrollViewer` but can also be used without.

![image](https://user-images.githubusercontent.com/20152272/86844315-31288800-c0a8-11ea-97f0-ed2018ed77c0.png)


## PR Type
What kind of change does this PR introduce?
- Feature

## What is the new behavior?
Add a new `MultiColumnsStackPanel`. It exposes the following properties:
- `ItemsSpacing`: the vertical space between the child elements
- `ColumnsSpacing`: the horizontal width between the columns
- `MaxColumnWidth`: the maximum width for the columns
_ `HorizontalContentAlignement`: the horizontal alignment for the children. If the child is a `Control` we will use its `Control.HorizontalAlignment` value instead.

It should still be possibe to improve the algorithm to reduce the number of computation. That's something that can be done later :)

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: <!-- docs PR link -->
- [x] Sample in sample app has been added / updated (for bug fixes / features)
    - [x] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes

michael-hawker commented:
Thanks @vgromfeld, this is an interesting scenario.

I'm trying to think of other panels we have already for this scenario like [WrapPanel](https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/wrappanel) w/ Orientation Vertical which can do some similar things (outside a ScrollViewer). I guess the difference here is it still tries to put things in columns explicitly? 

In which case, how does this differ from something like the [StaggeredPanel](https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/staggeredpanel)? Is it just the order of the layout, could we add an `Orientation` property to it instead?

If we move forward, I think we should align to other common spacing names like `RowSpacing` and `ColumnSpacing` or maybe `VerticalSpacing`?

FYI @azchohfi @skendrot @Sergio0694

skendrot commented:
This looks very much like the WrapPanel. However, when the Orientation is vertical and you have a ScrollViewer that scrolls vertically I don't think it will flow to another column. 
From what I can see, the advantage of this control is that the developers does not have to do the math to split the items into columns.

michael-hawker commented:
I don't think there's enough overlap in code to share with WrapPanel, but I wonder if we want a common interface for the properties that are shared like `HorizontalSpacing`, `VerticalSpacing`, and `Padding`?

vgromfeld commented:
> In which case, how does this differ from something like the [StaggeredPanel](https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/staggeredpanel)? Is it just the order of the layout, could we add an `Orientation` property to it instead?

The main difference is that `StaggeredPanel` will always try to have columns of equal height by adding items to the smallest columns. This implies that it can change the order of the items. When used in horizontal mode, it requires an horizontal scrollbar. 
In this panel, I want the first column(s) to always be fully filled and have the last one partially filled. It fully fills the columns one after the other. This panel always keep the children order.

> If we move forward, I think we should align to other common spacing names like `RowSpacing` and `ColumnSpacing` or maybe `VerticalSpacing`?

I agree, I will update the naming to stay aligned.

skendrot commented:
I wonder if we should add support for this in the WrapPanel instead of a new control. We could add a property for evenly distributing items

vgromfeld commented:
@skendrot, there is some overlap in how the two controls are displaying things but the approach is not the same. `MultiColumnsStackPanel` enforce the columns' width whereas `WrapPanel` does not and `MultiColumnsStackPanel` tries to find the best partition to limit scrolling. `WrapPanel` is more lightweight and only arrange the items one after the other.
It feels that merging the code of the two will end with a big *if/else* statement in the measure/arrange phases to run either one or the other layout. We should keep `WrapPanel` light (and faster).

vgromfeld commented:
@michael-hawker, I've renamed the vertical/horizontal spacing properties and added `Padding`.

michael-hawker commented:
@vgromfeld thanks, couple of StyleCop issues:

```
 "D:\a\1\s\Windows Community Toolkit.sln" (Build target) (1) ->
       "D:\a\1\s\Microsoft.Toolkit.Uwp.UI.Controls\Microsoft.Toolkit.Uwp.UI.Controls.csproj" (default target) (11:2) ->
       (CoreCompile target) -> 
         MultiColumnsStackPanel\MultiColumnsStackPanel.cs(363,5): error SA1508: A closing brace should not be preceded by a blank line. [D:\a\1\s\Microsoft.Toolkit.Uwp.UI.Controls\Microsoft.Toolkit.Uwp.UI.Controls.csproj]
         MultiColumnsStackPanel\MultiColumnsStackPanel.cs(53,1): error SA1507: Code should not contain multiple blank lines in a row [D:\a\1\s\Microsoft.Toolkit.Uwp.UI.Controls\Microsoft.Toolkit.Uwp.UI.Controls.csproj]
```

Any other thoughts on improvements or features? Is this pretty much done? Should we gather more feedback from the community? What approach would you like to take here in next steps?

michael-hawker commented:
@vgromfeld couple of stylecop things to fix still.

@RosarioPulella want to play around and see how things work a bit for any other feedback? If it looks good to you then I think we can just get this merged to have others try in the next preview.

vgromfeld commented:
@michael-hawker I've fixed the remaining stylecop issues.
I've also removed my `HorizontalContentAlignment` property since it is not needed (based on what we've done for the WrapPanel).
We should be good to go :)

RosarioPulella commented:
@vgromfeld I may be confused about some things, but if I understand correctly there are a few important behaviors to this control.

1. Fill the columns so their heights are as even as possible following a constraint in width.
2. Order the items from top to bottom of a collum and then from the left to the right column.
3. Fill up the first collum(s) and put the rest in the last column.

Using the first two behaviors this control can be looked at in two ways.

### `WrapPanel` + behavior 1 
Instead of using a constraint of space in the direction of orientation to dictate the creation of new row/columns along the direction of orientation. It fills the space perpendicular to the orientation with as many columns/rows as possible attempting to make them as equal in height/width as possible.

### `StaggeredPanel` + behavior 2
Instead of painting horizontally layers of elements on top of each other vertically. It adds elements from one column to the next but the right amount of elements in each column to make the columns as close in height as possible.

Is behavior 3 a consequence of the implementation, rather than a desired behavior?

Are the `WrapPenel` and `StaggeredPanel` perspectives correct?

michael-hawker commented:
Thanks @vgromfeld! I pulled down the update, looks like the 'Field 2' example isn't how you'd want it to be?

![image](https://user-images.githubusercontent.com/24302614/98601907-88c44380-2294-11eb-9a57-9ec2a1a50ea7.png)

I think you wanted these all in one line, no? Let me know if this is expected.

We should also provide some default padding for the example. Also, is there a good value we can use for the larger boxes so that the grid is more likely to match across a bit more?

![image](https://user-images.githubusercontent.com/24302614/98602365-2f104900-2295-11eb-994b-0507778eda89.png)

i.e. make it more likely for field 17 to fall inline across with field 6... (and that field 6 would then be tall enough to encompass 17 & 18). Think it's just a matter of finding the right multiple of the default height?

michael-hawker commented:
@RosarioPulella for comparison here are the current outputs with the other two controls. WrapPanel just keeps going and won't wrap. StaggeredPanel is certainly closer but goes left-to-right instead of top-to-bottom-across.

<details>
    <summary>WrapPanel XAML</summary>

```xml
    <ScrollViewer HorizontalScrollBarVisibility="Disabled"
                      HorizontalScrollMode="Disabled"
                      VerticalScrollBarVisibility="Auto"
                      VerticalScrollMode="Auto">
            <controls:WrapPanel HorizontalAlignment="Left"
                                             VerticalAlignment="Top"
                                             HorizontalSpacing="12"
                                             VerticalSpacing="8"
                                             Orientation="Vertical"
                                             Padding="0,0,0,0">
                <TextBox Header="Field 1" />
                <TextBox Header="Field 2" />
                <TextBox Height="200"
                         Header="Field 3" />
                <TextBox Header="Field 4" />
                <TextBox Header="Field 5" />
                <TextBox Height="100"
                         Header="Field 6" />
                <TextBox Height="200"
                         Header="Field 7" />
                <TextBox Header="Field 8" />
                <TextBox Height="100"
                         Header="Field 9" />
                <TextBox Header="Field 10" />
                <TextBox Header="Field 11" />
                <TextBox Header="Field 12" Height="64" />
                <TextBox Header="Field 13" />
                <TextBox Height="200"
                         Header="Field 14" />
                <TextBox Header="Field 15" />
                <TextBox Height="100"
                         Header="Field 16" />
                <TextBox Header="Field 17" />
                <TextBox Header="Field 18" />
                <TextBox Header="Field 19" />
                <TextBox Height="200"
                         Header="Field 20" />
                <TextBox Header="Field 21" />
                <TextBox Height="100"
                         Header="Field 22" />
                <TextBox Header="Field 23" />
                <TextBox Height="264"
                         Header="Field 24" />
                <TextBox Header="Field 25" />
                <TextBox Height="200"
                         Header="Field 26" />
                <TextBox Header="Field 27" />
                <TextBox Header="Field 28" />
                <TextBox Height="200"
                         Header="Field 29" />
                <TextBox Header="Field 30" />
            </controls:WrapPanel>
        </ScrollViewer>
```
</details>

![image](https://user-images.githubusercontent.com/24302614/98602798-e6a55b00-2295-11eb-8066-15e3bcaf5dc2.png)

<details>
    <summary>StaggeredPanel XAML</summary>

```xml
        <ScrollViewer HorizontalScrollBarVisibility="Disabled"
                      HorizontalScrollMode="Disabled"
                      VerticalScrollBarVisibility="Auto"
                      VerticalScrollMode="Auto">
            <controls:StaggeredPanel HorizontalAlignment="Left"
                                             VerticalAlignment="Top"
                                             DesiredColumnWidth="240"
                                             ColumnSpacing="12"
                                             RowSpacing="8"
                                             Padding="0,0,0,0">
                <TextBox Header="Field 1" />
                <TextBox Header="Field 2" />
                <TextBox Height="200"
                         Header="Field 3" />
                <TextBox Header="Field 4" />
                <TextBox Header="Field 5" />
                <TextBox Height="100"
                         Header="Field 6" />
                <TextBox Height="200"
                         Header="Field 7" />
                <TextBox Header="Field 8" />
                <TextBox Height="100"
                         Header="Field 9" />
                <TextBox Header="Field 10" />
                <TextBox Header="Field 11" />
                <TextBox Header="Field 12" Height="64" />
                <TextBox Header="Field 13" />
                <TextBox Height="200"
                         Header="Field 14" />
                <TextBox Header="Field 15" />
                <TextBox Height="100"
                         Header="Field 16" />
                <TextBox Header="Field 17" />
                <TextBox Header="Field 18" />
                <TextBox Header="Field 19" />
                <TextBox Height="200"
                         Header="Field 20" />
                <TextBox Header="Field 21" />
                <TextBox Height="100"
                         Header="Field 22" />
                <TextBox Header="Field 23" />
                <TextBox Height="264"
                         Header="Field 24" />
                <TextBox Header="Field 25" />
                <TextBox Height="200"
                         Header="Field 26" />
                <TextBox Header="Field 27" />
                <TextBox Header="Field 28" />
                <TextBox Height="200"
                         Header="Field 29" />
                <TextBox Header="Field 30" />
            </controls:StaggeredPanel>
        </ScrollViewer>
```
</details>

![image](https://user-images.githubusercontent.com/24302614/98603392-d0e46580-2296-11eb-8d09-c494d0b02b65.png)

@skendrot if StaggeredPanel supported orientation to layout down first vs. across would that be similar? Though I feel like the logic across vs. down is quite a different problem, but seems to make sense from a usability/flow perspective?

Thoughts?

RosarioPulella commented:
@vgromfeld if what you are trying to accomplish with this control is essentially the row/column ordering of the `WrapPanel` but more or less to fill the available space the way the `StaggeredPanel` does, then perhaps the API can be made more clear one of several ways. 

1. Changing the name of the control to something more specific to its behavior. 
From a visual standpoint both the `WrapPanel`, `StaggeredPanel`, and this new control are all some sort of "Multi-Column" `StackPanel`. 
Where the `StackPanel` does not adapt to spatial constraints, the `WrapPanel`, `StaggeredPanel`, and this new control can be viewed as different ways of dealing with spatial contraints on a collection using a dynamic component to the layout of the items in the collection. 
2. Adding functionality to an exsiting area of the API to enable this usecase. (maybe via `WrapPanel` or `StaggeredPanel`

vgromfeld commented:
@RosarioPulella, you are rigth about the constraint the control tries to follow:
- Fill the columns so their heights are as even as possible following a constraint in width.
- Order the items from top to bottom of a column and then from the left to the right column.
- Fill up the first column(s) and put the rest in the last column (this depends on the number of children we have and how we can lay out them).

I agree that there is some overlap between this new control and the existing `WrapPanel` and `StaggeredPanel` controls regarding how they are displaying things but the approach is not the same. 
- `WrapPanel `does not enforce the columns' width and should not. `WrapPanel` is mostly a `StackPanel` which can arrange items on a new line when it reaches the end of a line. It uses lightweight computations to get this layout and should keep is performances.
- `StaggeredPanel` may be a better fit but the actual implementation is closed to the `WrapPanel`'s one. It arranges items on a new line once it reaches the end of a line but instead of restarting at the begining, it will put the items in the smallest column. This breaks the children order since the first element on the second row may not be the one after the last one of the first row. Adding the `MultiColumnsStackPanel` implementation in it feels like adding a big `if/else` statement in the measure/arrange phases to run either one or the other layout.

I choose to write a new panel to keep the existing ones light and simple.

I do agree that we can find a better name. I didn't spend a lot of time on the one I choose but didn't find anything better that `MultiColumnStackPanel` 😊.


@michael-hawker, I updated the sample code to add some `HorizontalAlignment` properties on the first children to validate the behavior but I can remove them since they can be confusing. I will update the heights of the text blocks to have a better alignment.

michael-hawker commented:
Thanks @vgromfeld, I think that's been the sticking point is it just feels so close even though the implementations are different. 🙂 I agree just adding a big if/else isn't going to be helpful. Let's keep it separate as you have it in this PR for now.

Good to know the buttons were just a test, I just wasn't sure what the desired behavior was as it seemed like a case where a dev might expect them to show up in the same positioning with the different alignments.

For naming, wondering if we want to call this `FormPanel`? It could actually be a good base if we want to tackle adding a prototype for the WinUI Form Control https://github.com/microsoft/microsoft-ui-xaml/issues/82 proposal in the future. Not sure if that would require more functionality for this panel or not (for the 'groups'), but we could investigate that later.

![image](https://user-images.githubusercontent.com/24302614/98932911-d3141300-2494-11eb-88f7-dd564e116d07.png)

It's something I could use in XAML Studio for my settings page too, so maybe it's something I can play with and prototype on one of my streams later... 🤔

RosarioPulella commented:
@vgromfeld I Hope I was not to hash on the name you chose. But I also think it makes a lot of sense as its own control. 

One thing I don't understand is why to try to fill the fist columns and leave the rest in the last? If we are trying to make all the other columns as close to each other in height as possible, then why to have one overflow-ish one that we just dump items into. It sounds like it would be more consistent if we just try to make all the columns around the same height.

I might not fully understand the driving use case, but is the point of this control to dynamically layout items into columns (near in height to each other) according to a constraint in width? I think that behavior alone is clear to users, but adding the bit about the last column seems inconsistent to me.

On a similar note, if the `StaggeredPanel` did something like that I would be confused about the reasoning behind it. It seems like it would be less general, more opinionated, and maybe useful to fewer users.

vgromfeld commented:
@michael-hawker , I like the `FormPanel` name. I've updated the code to use this name 😊. I've also removed the text controls and updated the height of all the text fields:

![image](https://user-images.githubusercontent.com/20152272/99317877-07a61700-2867-11eb-8b89-aa4279e01bb9.png)

@RosarioPulella the last/overflow column is mostly an artefact of the implementation. I probably need to improve the logic to have better columns' sizes. For now, the first column is driving everything and sometimes, based on the children heights, I might not get a fully filled last column. I will check if there is something easily doable.

RosarioPulella commented:
Thanks, @vgromfeld. That clears up the intended behavior and I am excited to see this control merged 😃! 
Also not sure if this behavior is intended, but it seems to adapt to the width of its container until 4 columns and stops distributing items past that.
![Form_fill](https://user-images.githubusercontent.com/7164552/99321687-699d5700-283c-11eb-93d3-c1dc9382c56b.gif)

Is it supposed to keep making columns like the `StaggeredPanel`?

![Staggered_fill](https://user-images.githubusercontent.com/7164552/99321890-df092780-283c-11eb-9861-eb617e54b2bf.gif)

vgromfeld commented:
@RosarioPulella, this is intended. The control tries to fill the full height of the container. If the container becomes so wide that we have more space than needed, only the "first" columns will be filled.

skendrot commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/FormPanel/FormPanelCode.bind:
```diff
@@ -0,0 +1,64 @@
+                                             VerticalSpacing="@[VerticalSpacing:Slider:8:0-64]"
+                                             MaxColumnWidth="@[MaxColumnWidth:Slider:240:0-500]"
+                                             Padding="@[Padding:Thickness:0,0,0,0]@">
+                <TextBox Header="Field 1" />
```
I think it would be best to generate these in code with random heights and allow the user to add new textboxes. Would be better for testing.

skendrot commented:
Doing some testing with adding random new fields with random heights and getting some behavior that seems off. Shouldn't these columns be closer in height? With the small heights of some of the boxes they could be shifted over.
![image](https://user-images.githubusercontent.com/1594689/99428723-3e5e4a80-28c4-11eb-8db1-df38e20423db.png)

michael-hawker commented:
Thanks for taking a look @skendrot, always appreciate your insights in this space!

@vgromfeld let us know when you have a chance to get back to looking at feedback, I know you're a bit busy. If you're in no rush too, I could see this spilling over into a 7.1 release so we can coordinate with the WTS folks in the comment I left in #3569 and test this out more with some more form helpers to complete the WinUI story vision... Thoughts?

michael-hawker commented:
Welcome to the Windows Community Toolkit @Loser09sly! We always love to see community folks reviewing features that are being worked on in the Toolkit.

Since this is your first interaction with our community, would you mind providing a few more details on what you tried for your review and your first impressions?

Thanks a bunch! Look forward to working with you more in the future.

michael-hawker commented:
@vgromfeld thoughts on where we are and what's left to do based on the feedback in the PR? Should we try and get this into 7.0 or hold back and evaluate adding more helpers for a larger story to support https://github.com/microsoft/microsoft-ui-xaml/issues/82 for 7.1 more fully and use that as a test-case for this panel?

vgromfeld commented:
@michael-hawker, we should wait for 7.1 and see how we can integrate some of the form control elements in it.

Kyaa-dost commented:
@vgromfeld Thanks for the input. 

@michael-hawker will add this in 7.1 for now and then will proceed from there.
